torch/nn/modules/module.py

```diff
@@ -2333,7 +2333,14 @@ def zero_grad(self, set_to_none: bool = False) -> None:
                         p.grad.detach_()
                     else:
                         p.grad.requires_grad_(False)
-                    p.grad.zero_()
+                    if p.grad.device.type == 'cuda':
+                        per_device_and_dtype_grads[p.grad.device][p.grad.dtype].append(p)
+                    else:
+                        p.grad.zero_()
+        if per_device_and_dtype_grads:
+            for _, per_dtype_grads in per_device_and_dtype_grads.items():
```

Comment:
`for per_dtype_grads in per_device_and_dtype_grads.values()`

Reply:
Btw, shouldn't foreach methods do such grouping internally? Because this grouping seems a frequent idiom now...

Reply:
> Btw, shouldn't foreach methods do such grouping internally? Because this grouping seems a frequent idiom now...

I think it is supposed to be. But for now, only CUDA supports `foreach` , so that's why we need to check the device name. I assume the next step is to eliminate the foreach argument in this function. But for now, need some work to make sure this kind of check is correct.

Reply:
We had a few discussions on this. The short answer is that yes they could, but there are also many cases where we prefer to do it before.
For example for the optimizers, we prefer to do the grouping once for the whole step() function.
So it is not clear that we want each function to do it. For now using a common util to do this grouping is a good compromise I think.